staging/src/k8s.io/component-base/config/testing/helpers.go

```diff
@@ -28,69 +29,63 @@ import (
-			if !ok {
-				t.Errorf("unable to locate encoder -- %q is not a supported media type", mediaType)
-			}
+// TestCase defines a testcase for roundtrip and defaulting tests
+type TestCase struct {
+	name, in, out string
+	codec         runtime.Codec
+}
 
-			encoder := codecs.EncoderForVersion(info.Serializer, rt.outGV)
+func decodeYAML(t *testing.T, path string, codec runtime.Codec) runtime.Object {
+	content, err := ioutil.ReadFile(path)
+	if err != nil {
+		t.Fatal(err)
+	}
 
-			actual, err := runtime.Encode(encoder, obj)
-			if err != nil {
-				t.Fatalf("failed to encode object: %v", err)
-			}
+	// decode to internal type
+	object, err := runtime.Decode(codec, content)
+	if err != nil {
+		t.Fatal(err)
+	}
 
-			expected, err := ioutil.ReadFile(rt.out)
-			if err != nil && !os.IsNotExist(err) {
-				t.Fatalf("couldn't read test data: %v", err)
-			}
+	return object
+}
 
-			needsUpdate := false
-			if os.IsNotExist(err) {
-				needsUpdate = true
-				t.Error("couldn't find test data")
-			} else {
-				if !bytes.Equal(expected, actual) {
-					t.Errorf("Output does not match expected, diff (- want, + got):\n\tin: %s\n\tout: %s\n\tgroupversion: %s\n\tdiff: \n%s\n",
-						rt.in, rt.out, rt.outGV.String(), cmp.Diff(string(expected), string(actual)))
-					needsUpdate = true
-				}
-			}
-			if needsUpdate {
-				const updateEnvVar = "UPDATE_COMPONENTCONFIG_FIXTURE_DATA"
-				if os.Getenv(updateEnvVar) == "true" {
-					if err := ioutil.WriteFile(rt.out, actual, 0755); err != nil {
-						t.Fatal(err)
-					}
-					t.Logf("wrote expected test data... verify, commit, and rerun tests")
-				} else {
-					t.Logf("if the diff is expected because of a new type or a new field, re-run with %s=true to update the compatibility data", updateEnvVar)
-				}
-			}
-		})
+func getCodecForGV(codecs serializer.CodecFactory, gv schema.GroupVersion) (runtime.Codec, error) {
+	mediaType := runtime.ContentTypeYAML
+	serializerInfo, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), mediaType)
+	if !ok {
+		return nil, fmt.Errorf("unable to locate encoder -- %q is not a supported media type", mediaType)
 	}
+	codec := codecs.CodecForVersions(serializerInfo.Serializer, codecs.UniversalDeserializer(), gv, nil)
+	return codec, nil
 }
 
-func decodeTestData(path string, scheme *runtime.Scheme, gvk schema.GroupVersionKind, codecs serializer.CodecFactory) (runtime.Object, error) {
-	content, err := ioutil.ReadFile(path)
-	if err != nil {
-		return nil, err
+func matchOutputFile(t *testing.T, outputFile string, actual []byte) {
+	expected, err := ioutil.ReadFile(outputFile)
+	if err != nil && !os.IsNotExist(err) {
+		t.Fatalf("couldn't read test data: %v", err)
 	}
 
-	obj, _, err := codecs.DecoderToVersion(codecs.UniversalDecoder(), gvk.GroupVersion()).Decode(content, &gvk, nil)
-	if err != nil {
-		return nil, err
+	needsUpdate := false
+	if os.IsNotExist(err) {
+		needsUpdate = true
+		t.Error("couldn't find test data")
```

Comment:
updating here will cause printing this message even if the env var is set. I've the last message to this `if the diff is expected because of a new type or a new field, "re-run with  UPDATE_COMPONENTCONFIG_FIXTURE_DATA=true to update the compatibility data or generate missing files`